### ISSUE TITLE
Prevent multiple clicks/presses on the same cell

### ISSUE THREAD

**Issue by arasaac-dga** [asterics/AsTeRICS-Grid#169]
Prevent the accumulation of the same symbols several times to the accumulated phrase when the user presses it repeatedly (to prevent multiple clicks on the same cell). It must be an option in "user settings" that check/uncheck in case neccesary. When is active, only once click/press is allowed (and the cell remain "blocked") until the user press another cell.

**Comment by arasaac-dga** [comment]
As addtional information about the need of this option we have observed that repeatedly clicking on the same cell causes again the problem of duplicity of the accumulated sentence that was solved in other issue.

**Comment by anlolli92** [comment]
I also would find this feature very useful! If you could please implement...

Thank you!

**Comment by klues** [comment]
I assume that the only thing to block multiple times in a row is adding the same thing multiple times to a collection element, correct?
I think for other actions it makes no sense to block them to be performed multiple times, but it would lead to unwanted behaviour if e.g. duplicated "navigate back" or "play/pause video" wouldn't be allowed.

**Comment by klues** [comment]
Therefore I maybe would add this option as part of the collect element options.

**Comment by klues** [comment]
Please test on https://grid.asterics.eu/latest/

1. Edit global grid:
![grafik](https://user-images.githubusercontent.com/2537025/199700386-60eab480-a9ae-4a6f-bc99-538564953f92.png)

2. Check "Prevent collecting the same cell several times in a row":
![grafik](https://user-images.githubusercontent.com/2537025/199700572-abbeb4ce-c5da-4cd1-bfd2-a01388f55d0b.png)

**Comment by arasaac-dga** [comment]
We have test it and works as expected. Perhaps only remain to confirm that this option solve the problem of duplicity of the accumulated sentence that we have commented.

**Comment by klues** [comment]
Thanks for the hint, I've now also fixed the duplicity of the accumulated sentence when pressing a navigation element multiple times, please test on grid.asterics.eu/latest/
It's independent of the other option.

**Comment by arasaac-dga** [comment]
​Ok, thanks. We think that linked cells don't create problems because once pressed it goes to the other grid and only is added one time to the accumulated phrase. But if it's controlled (in case of slow tablets), sure it will be better. Is there an option to activate/deactivate this second option? or is the same?​

**Comment by ChrisVeigl** [comment]
just one idea: could we add an option for "anti-tremor" which allows users to define a "minimum touch time" for a cell in order to activate it, and an "inactivity timespan" where any additional touch will be ignored after a cell was activated ?

**Comment by klues** [comment]
> ​Ok, thanks. We think that linked cells don't create problems because once pressed it goes to the other grid and only is added one time to the accumulated phrase. But if it's controlled (in case of slow tablets), sure it will be better. Is there an option to activate/deactivate this second option? or is the same?​

What I've implemented now are 2 things:
* The possibility to prevent adding the same elements two or more times in a row to the collection element -> can be set as option on the collect element. Default is disabled, since it also prevents 2 times the same letter on a keyboard grid.
* The issue with duplicated global grids when selecting a navigation element 2 times was fixed independently of the option above, so it should always be fixed.

> just one idea: could we add an option for "anti-tremor" which allows users to define a "minimum touch time" for a cell in order to activate it, and an "inactivity timespan" where any additional touch will be ignored after a cell was activated ?

I think "minimum touch time" is already possible using the input method "Hovering". Inactivity timespan would be possible, however I think I only would add it if it's requested by a user since we don't know if it's really needed/used. The main problem were duplicated entries in the collect element, which is now resolved with the new option. Otherwise multiple presses on the same element aren't a problem in most cases since e.g. speaking the text 2 times won't be that bad since the second speech will interrupt the first one. And in general it's questionable what input method people with tremor problems will use, probably not clicking. And for everything with button input, there are already "anti-tremor-options", see https://www.asterics.eu/manuals/asterics-grid/04_input_options.html#introduction-to-input-events

**Comment by arasaac-dga** [comment]
As you comment, Benjamin, the "anti-tremor" option could be solved with "Hovering" option. The option to set a "inactivity timespan" (delay) over the last cell pressed, we consider that it's useful because in our experience, for many children with autism the feedback (auditive) that generate when press multiple times the same cell (e.g wa-wa-wa-want) is a reinforcement for them because like to listen it. Other communicators have this option. For us, the best will be that "inactivity timespan" acts only over the last press cell and with a configurable time (2-3 seconds by default). On the other hand, we consider that it must be an user option not activated by default.

**Comment by klues** [comment]
OK, for "inactivity timespan" again there is the question where to apply it?
If it prevents all actions of an element in the timespan it's also not possible anymore to start/stop a video or radio or to navigate back two times in a row. What I supposing now is that it's only needed for duplicated voice messages?! And possibly then it's not necessary to have 2 different options for voice and for collect elements, but only a single one like `Minimum pause for collecting and speaking the same cell several times in a row` where the user can select a time.

What do you think?

**Comment by arasaac-dga** [comment]
Yes, we agree with your proposal to add only one option "Minimum pause for collecting and speaking the same cell several times in a row" with the possibility to set a time for delay (perhaps between 0,5 seconds and 10). This option must be deactivated by default and activated if the user needs it. The main problem once activated can be with the use of a keyboard (when you write a word with a repeated letter) but, setting the delay time, we understand could be adjusted to the user. If the user doesn't use the keyboard the delay time could be higher than if the user uses the keyboard. 

On the other hand, with the use of this months and the comments of the users, we consider than the General Configuration options of the accumulated phrase must be in Configuration page too (duplicate them or move to this page) because it's difficult to the users find them and there are very important and interesting. So study the possibility of moving or duplicating all the general options (not the actions) of the accumulated phrase to the Configuration page to improve usability.

**Comment by anlolli92** [comment]
Hi @benjaminaigner 

We are trying to test it in https://grid.asterics.eu/latest/ , but no luck! If we go to Edit / More / Edit global grid, we cannot see these options, but only blank grid:

![Captura de pantalla 2022-11-05 a las 9 22 21](https://user-images.githubusercontent.com/116308201/200110724-607b164d-181b-430c-b57b-8300f8234bb2.png)

![Captura de pantalla 2022-11-05 a las 9 22 30](https://user-images.githubusercontent.com/116308201/200110735-4b639683-aced-4da2-98b9-beac7c0ec9d2.png)

Maybe this "latest" is not the right version to try this anymore? Any clue?

Thank you very much!

**Comment by klues** [comment]
I've just released a new version, please test https://grid.asterics.eu/latest/ again. I've removed the special option from to collect element and added this new option to the settings page:

![grafik](https://user-images.githubusercontent.com/2537025/200347669-9b9de1f8-162d-4304-9908-b1b925657807.png)

Along with the options for acoustic feedback these options also can be set within the settings of the input options: 

![grafik](https://user-images.githubusercontent.com/2537025/200347982-52c2efb5-e5fa-4c40-9815-6b85c313a1b1.png)

Please test and tell me your feedback.

> On the other hand, with the use of this months and the comments of the users, we consider than the General Configuration options of the accumulated phrase must be in Configuration page too

Please open a separate issue for this.

**Comment by anlolli92** [comment]
Hi @benjaminaigner ,

Just tested, and works as expected! Many thanks!

**Comment by klues** [comment]
Thanks for testing, but I'm @klues while @benjaminaigner is my colleague here at UAS Technikum Wien, who isn't coding on the AsTeRICS Grid project ;)
I'll wait for the feedback from ARASAAC and then I'll release it to the main version.

**Comment by arasaac-dga** [comment]
Tested and works as expected. Great job!! In #180 we have made a proposal of settings page organization with this option included.

**Comment by klues** [comment]
Released with https://github.com/asterics/AsTeRICS-Grid/releases/tag/release-2022-11-10-10.40%2F%2B0100
